flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -249,6 +250,22 @@ public void testPartitionPromotion() throws Exception {
 		);
 	}
 
+	@Test
+	public void testClusterPartitionRelease() throws Exception {
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutor, taskExecutorGateway, partitionTracker) -> {
+				final CompletableFuture<Collection<IntermediateDataSetID>> releasePartitionsFuture = new CompletableFuture<>();
+				runInTaskExecutorThreadAndWait(taskExecutor, () -> partitionTracker.setReleaseClusterPartitionsConsumer(releasePartitionsFuture::complete));
+
+				final IntermediateDataSetID dataSetId = new IntermediateDataSetID();
+
+				taskExecutorGateway.releaseClusterPartitions(Collections.singleton(dataSetId));
+
+				assertThat(releasePartitionsFuture.get(), hasItems(dataSetId));
+			}
+		);
+	}
```

Comment:
I'm not entirely sure that this test is really necessary. What we are effectively testing here is that the `TaskExecutor` forwards the `releaseClusterPartitions` to the `partitionTracker`. For that to test, we do quite some extra work in `testPartitionRelease`. I think if we only want to test the call forwarding, then we don't have to do the extra work done by `testPartitionRelease`. If we want to test the integration of the `TaskExecutor` and the `TaskExecutorPartitionTracker`, then we might consider writing an integration test which uses the proper implementation. Of course, then we need to choose a different way to verify the correct behaviour (different level of mock components).

Reply:
While it does do a fair amount of setup work it does verify the expected behavior for the _expected circumstances_; a partition release after a task has finished.
The whole point of `testPartitionRelease` is to create a "realistic" environment; remove any part of it and we are working in a theoretical scenario that may (now or at some point) no longer represent reality.

We could streamline the test a bit such that we test all things at once (so we only pay the setup cost once), but that would be out-of-scope of this PR.

At the same time the current test(s) have the usual problem of our mocking-like approach; they check that something was called, but other interactions are fully ignored. Take `testPartitionPromotion` as an example; for all we know the TE may (wrongly) release all partitions after the promotion, and the test wouldn't cover that.
I would prefer if the tests were to use the actual implementation, but I don't know of a good way to wait for the TM to have fully processed a request. I remember that we were looking to a solution for this some time ago, but don't remember what the conclusion was.